In [1]:
import pandas as pd

In [2]:
data = {"apples": [3,2,0,1],
        "oranges":[0,3,7,2]
       }

In [5]:
purchases = pd.DataFrame(data)
purchases

apples  oranages
0       3         0
1       2         3
2       0         7
3       1         2

Each (key, value) item in data corresponds to a column in the resulting DataFrame.

The Index of this DataFrame was given to us on creation as the numbers 0-3, but we could also create our own when we initialize the DataFrame.

Let's have customer names as our index:

In [7]:
purchases = pd.DataFrame(data, index = ["Johnny", "Sravs", "Lilly", "Robert"])
purchases

apples  oranages
Johnny       3         0
Sravs        2         3
Lilly        0         7
Robert       1         2

So now we could locate a customer's order by using their name:

In [8]:
purchases.loc["Sravs"]

apples      2
oranages    3
Name: Sravs, dtype: int64

There's more on locating and extracting data from the DataFrame later, but now you should be able to create a DataFrame with any random data to learn on.

Let's move on to some quick methods for creating DataFrames from various other sources.

### How to read in data?

In [10]:
df = pd.read_csv("C:\\Users\\johny\\Desktop\\purchases.csv")
df

Unnamed: 0  apples  oranges
0       June       3        0
1     Robert       2        3
2       Lily       0        7
3      David       1        2

CSVs don't have indexes like our DataFrames, so all we need to do is just designate the index_col when reading:

In [24]:
#Here we're setting the index to be column zero.

df =  pd.read_csv("C:\\Users\\johny\\Desktop\\purchases.csv", index_col = 0)
df

apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2

You'll find that most CSVs won't ever have an index column and so usually you don't have to worry about this step.

## Most important DataFrame operations

Let's load in the IMDB movies dataset to begin:

In [25]:
movies_df = pd.read_csv("C:\\Users\\johny\\Desktop\\IMDB-Movie-Data.csv")
movies_df. head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  
0     8.1  757074              333.13       76.0  
1     7.0  485820              126.46       65.0  
2     7.3  157606              138.12       62.0  
3     7.2   60545              270.32       59.0  
4     6.2  393727              325.02       40.0

In [27]:
#Designating the movie titles to be our index
movies_df = pd.read_csv("C:\\Users\\johny\\Desktop\\IMDB-Movie-Data.csv", index_col = "Title")
movies_df.head()

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
Split                                138.12       62.0  
Sing                                 270.32       59.0  
Suicide Squad                        325.02       40.0

We designated the movie titles to be our index

In [28]:
movies_df.tail(2)

Rank                  Genre  \
Title                                       
Search Party   999       Adventure,Comedy   
Nine Lives    1000  Comedy,Family,Fantasy   

                                                    Description  \
Title                                                             
Search Party  A pair of friends embark on a mission to reuni...   
Nine Lives    A stuffy businessman finds himself trapped ins...   

                      Director  \
Title                            
Search Party    Scot Armstrong   
Nine Lives    Barry Sonnenfeld   

                                                         Actors  Year  \
Title                                                                   
Search Party  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
Nine Lives    Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

              Runtime (Minutes)  Rating  Votes  Revenue (Millions)  Metascore  
Title                                                                          
Search Party                 93     5.6   4881                 NaN       22.0  
Nine Lives                   87     5.3  12435               19.64       11.0

### Getting info about your data

In [29]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, Guardians of the Galaxy to Nine Lives
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Genre               1000 non-null   object 
 2   Description         1000 non-null   object 
 3   Director            1000 non-null   object 
 4   Actors              1000 non-null   object 
 5   Year                1000 non-null   int64  
 6   Runtime (Minutes)   1000 non-null   int64  
 7   Rating              1000 non-null   float64
 8   Votes               1000 non-null   int64  
 9   Revenue (Millions)  872 non-null    float64
 10  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(4)
memory usage: 93.8+ KB


__.info()__ provides the essential details about your dataset, such as the number of rows and columns, the number of non-null values, what type of data is in each column, and how much memory your DataFrame is using.

In [31]:
movies_df.shape

(1000, 11)

Note that .shape has no parentheses and is a simple tuple of format (rows, columns). 

### Handling duplicates

This dataset does not have duplicate rows, but it is always important to verify you aren't aggregating duplicate rows.

To demonstrate, let's simply just double up our movies DataFrame by appending it to itself:

In [34]:
temp_df = movies_df.append(movies_df)
temp_df.shape

(2000, 11)

Using **append()** will return a copy without affecting the original DataFrame. We are capturing this copy in temp so we aren't working with the real data.

Notice call __.shape__ quickly proves our DataFrame rows have doubled.

Now we can try dropping duplicates:

In [35]:
temp = temp_df.drop_duplicates()
temp.shape

(1000, 11)

Just like __append()__, the __drop_duplicates()__ method will also return a copy of your DataFrame, but this time with duplicates removed. Calling __.shape__ confirms we're back to the 1000 rows of our original dataset.

It's a little verbose to keep assigning DataFrames to the same variable like in this example. For this reason, pandas has the __inplace__ keyword argument on many of its methods. Using __inplace=True__ will modify the DataFrame object in place:

In [36]:
temp_df.drop_duplicates(inplace = True)

In [37]:
temp.shape

(1000, 11)

Now our __temp_df__ will have the transformed data automatically.

Another important argument for __drop_duplicates()__ is `keep`, which has three possible options:

- `first`: (default) Drop duplicates except for the first occurrence.
- `last`: Drop duplicates except for the last occurrence.
- `False`: Drop all duplicates.

Since we didn't define the `keep` arugment in the previous example it was defaulted to `first`. This means that if two rows are the same pandas will drop the second row and keep the first row. Using `last` has the opposite effect: the first row is dropped.

`keep`, on the other hand, will drop all duplicates. If two rows are the same then both will be dropped. Watch what happens to temp_df:

In [41]:
temp_df = movies_df.append(movies_df)
temp_df.drop_duplicates(inplace = True, keep = False)
temp_df.shape

(0, 11)

Since all rows were duplicates, keep=False dropped them all resulting in zero rows being left over. If you're wondering why you would want to do this, one reason is that it allows you to locate all duplicates in your dataset. When conditional selections are shown below you'll see how to do that.

### Column cleanup

Many times datasets will have verbose column names with symbols, upper and lowercase words, spaces, and typos. To make selecting data by column name easier we can spend a little time cleaning up their names.

Here's how to print the column names of our dataset:

In [42]:
movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

Not only does .columns come in handy if you want to rename columns by allowing for simple copy and paste, it's also useful if you need to understand why you are receiving a Key Error when selecting data by column.

We can use the __.rename()__ method to rename certain or all columns via a __dict__. We don't want parentheses, so let's rename those:

In [48]:
movies_df.rename(columns = {"Runtime (Minutes)": 'Runtime', "Revenue (Millions)": 'Revenue_millions'}, inplace = True)
movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year', 'Runtime',
       'Rating', 'Votes', 'Revenue_millions', 'Metascore'],
      dtype='object')

Excellent. But what if we want to lowercase all names? Instead of using .rename() we could also set a list of names to the columns like so:

In [49]:
movies_df.columns = ['rank', "genre", 'description', 'director', 'actors', 'year', 'runtime', 'rating', 'votes', 'revenue_millions', 'metascore']
movies_df.columns

Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

But that's too much work. Instead of just renaming each column manually we can do a list comprehension:

In [50]:
movies_df.columns = [col.lower() for col in movies_df]
movies_df.columns


Index(['rank', 'genre', 'description', 'director', 'actors', 'year', 'runtime',
       'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

### How to work with missing values

When exploring data, you’ll most likely encounter missing or null values, which are essentially placeholders for non-existent values. Most commonly you'll see Python's None or NumPy's np.nan, each of which are handled differently in some situations.

There are two options in dealing with nulls:

1. Get rid of rows or columns with nulls
2. Replace nulls with non-null values, a technique known as imputation

Let's calculate to total number of nulls in each column of our dataset. The first step is to check which cells in our DataFrame are null:

In [130]:
movies_df.isnull()

rank  genre  description  director  actors   year  \
Title                                                                         
Guardians of the Galaxy  False  False        False     False   False  False   
Prometheus               False  False        False     False   False  False   
Split                    False  False        False     False   False  False   
Sing                     False  False        False     False   False  False   
Suicide Squad            False  False        False     False   False  False   
...                        ...    ...          ...       ...     ...    ...   
Secret in Their Eyes     False  False        False     False   False  False   
Hostel: Part II          False  False        False     False   False  False   
Step Up 2: The Streets   False  False        False     False   False  False   
Search Party             False  False        False     False   False  False   
Nine Lives               False  False        False     False   False  False   

                         runtime  rating  votes  revenue_millions  metascore  \
Title                                                                          
Guardians of the Galaxy    False   False  False             False      False   
Prometheus                 False   False  False             False      False   
Split                      False   False  False             False      False   
Sing                       False   False  False             False      False   
Suicide Squad              False   False  False             False      False   
...                          ...     ...    ...               ...        ...   
Secret in Their Eyes       False   False  False             False      False   
Hostel: Part II            False   False  False             False      False   
Step Up 2: The Streets     False   False  False             False      False   
Search Party               False   False  False             False      False   
Nine Lives                 False   False  False             False      False   

                         rating_category  
Title                                     
Guardians of the Galaxy            False  
Prometheus                         False  
Split                              False  
Sing                               False  
Suicide Squad                      False  
...                                  ...  
Secret in Their Eyes               False  
Hostel: Part II                    False  
Step Up 2: The Streets             False  
Search Party                       False  
Nine Lives                         False  

[1000 rows x 12 columns]

In [52]:
#To count the number of nulls in each column we use an aggregate function for summing
movies_df.isnull().sum()

rank                  0
genre                 0
description           0
director              0
actors                0
year                  0
runtime               0
rating                0
votes                 0
revenue_millions    128
metascore            64
dtype: int64

__Removing null values__

Data Scientists and Analysts regularly face the dilemma of dropping or imputing null values, and is a decision that requires intimate knowledge of your data and its context. Overall, removing null data is only suggested if you have a small amount of missing data.

In [53]:
#This operation will delete any row with at least a single null value

#movies_df.dropna()


Other than just dropping rows, you can also drop columns with null values by setting _axis=1_

In [55]:
#movies_df.dropna(axis = 1)

__Imputation__

Imputation is a conventional feature engineering technique used to keep valuable data that have null values.

There may be instances where dropping every row with a null value removes too big a chunk from your dataset, so instead we can impute that null with another value, usually the mean or the median of that column.

Let's look at imputing the missing values in the revenue_millions column. First we'll extract that column into its own variable:

In [56]:
revenue = movies_df["revenue_millions"]

Using square brackets is the general way we select columns in a DataFrame.



If you remember back to when we created DataFrames from scratch, the keys of the `dict` ended up as column names. Now when we select columns of a DataFrame, we use brackets just like if we were accessing a Python dictionary.

_revenue_ now contains a Series:

In [58]:
revenue.head()

Title
Guardians of the Galaxy    333.13
Prometheus                 126.46
Split                      138.12
Sing                       270.32
Suicide Squad              325.02
Name: revenue_millions, dtype: float64

Slightly different formatting than a DataFrame, but we still have our Title index.

We'll impute the missing values of revenue using the mean. Here's the mean value:

In [62]:
#revenue_mean = movies_df["revenue_millions"].mean()
#revenue_mean

In [61]:
revenue_mean = revenue.mean()
revenue_mean

82.95637614678897

With the mean, let's fill the nulls using __fillna()__

In [67]:
revenue.fillna(revenue_mean, inplace = True)

We have now replaced all nulls in revenue with the mean of the column. Notice that by using __inplace=True__ we have actually affected the original movies_df:

In [68]:
movies_df.isnull().sum()

rank                 0
genre                0
description          0
director             0
actors               0
year                 0
runtime              0
rating               0
votes                0
revenue_millions     0
metascore           64
dtype: int64

Imputing an entire column with the same value like this is a basic example. It would be a better idea to try a more granular imputation by Genre or Director.

For example, you would find the mean of the revenue generated in each genre individually and impute the nulls in each genre with that genre's mean.

Let's now look at more ways to examine and understand the dataset.

__Understanding your variables__

Using __describe()__ on an entire DataFrame we can get a summary of the distribution of continuous variables:

In [70]:
movies_df.describe()

rank         year      runtime       rating         votes  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000   113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962    18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000    66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000   100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000   111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000   123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000   191.000000     9.000000  1.791916e+06   

       revenue_millions   metascore  
count       1000.000000  936.000000  
mean          82.956376   58.985043  
std           96.412043   17.194757  
min            0.000000   11.000000  
25%           17.442500   47.000000  
50%           60.375000   59.500000  
75%           99.177500   72.000000  
max          936.630000  100.000000

Understanding which numbers are continuous also comes in handy when thinking about the type of plot to use to represent your data visually.

`.describe()` can also be used on a categorical variable to get the count of rows, unique count of categories, top category, and freq of top category:

In [71]:
movies_df["genre"].describe()

count                        1000
unique                        207
top       Action,Adventure,Sci-Fi
freq                           50
Name: genre, dtype: object

This tells us that the genre column has 207 unique values, the top value is Action/Adventure/Sci-Fi, which shows up 50 times (freq).

__.value_counts()__ can tell us the frequency of all values in a column:

In [73]:
movies_df["genre"].value_counts().head(10)

Action,Adventure,Sci-Fi       50
Drama                         48
Comedy,Drama,Romance          35
Comedy                        32
Drama,Romance                 31
Action,Adventure,Fantasy      27
Animation,Adventure,Comedy    27
Comedy,Drama                  27
Comedy,Romance                26
Crime,Drama,Thriller          24
Name: genre, dtype: int64

__Relationships between continuous variables__

By using the correlation method .corr() we can generate the relationship between each continuous variable:

In [74]:
movies_df.corr()

rank      year   runtime    rating     votes  \
rank              1.000000 -0.261605 -0.221739 -0.219555 -0.283876   
year             -0.261605  1.000000 -0.164900 -0.211219 -0.411904   
runtime          -0.221739 -0.164900  1.000000  0.392214  0.407062   
rating           -0.219555 -0.211219  0.392214  1.000000  0.511537   
votes            -0.283876 -0.411904  0.407062  0.511537  1.000000   
revenue_millions -0.252996 -0.117562  0.247834  0.189527  0.607941   
metascore        -0.191869 -0.079305  0.211978  0.631897  0.325684   

                  revenue_millions  metascore  
rank                     -0.252996  -0.191869  
year                     -0.117562  -0.079305  
runtime                   0.247834   0.211978  
rating                    0.189527   0.631897  
votes                     0.607941   0.325684  
revenue_millions          1.000000   0.133328  
metascore                 0.133328   1.000000

Correlation tables are a numerical representation of the bivariate relationships in the dataset.

Positive numbers indicate a positive correlation — one goes up the other goes up — and negative numbers represent an inverse correlation — one goes up the other goes down. 1.0 indicates a perfect correlation.

So looking in the first row, first column we see __rank__ has a perfect correlation with itself, which is obvious. On the other hand, the correlation between __votes__ and __revenue_millions__ is 0.6. A little more interesting.

Examining bivariate relationships comes in handy when you have an outcome or dependent variable in mind and would like to see the features most correlated to the increase or decrease of the outcome. You can visually represent bivariate relationships with scatterplots.

### DataFrame slicing, selecting, extracting

It's important to note that, although many methods are the same, DataFrames and Series have different attributes, so you'll need be sure to know which type you are working with or else you will receive attribute errors.

Let's look at working with columns first.

__By Columns__



In [80]:
#to extract a column using square brackets 
genre_col = movies_df["genre"]
genre_col

Title
Guardians of the Galaxy     Action,Adventure,Sci-Fi
Prometheus                 Adventure,Mystery,Sci-Fi
Split                               Horror,Thriller
Sing                        Animation,Comedy,Family
Suicide Squad              Action,Adventure,Fantasy
                                     ...           
Secret in Their Eyes            Crime,Drama,Mystery
Hostel: Part II                              Horror
Step Up 2: The Streets          Drama,Music,Romance
Search Party                       Adventure,Comedy
Nine Lives                    Comedy,Family,Fantasy
Name: genre, Length: 1000, dtype: object

In [79]:
type(genre_column)

pandas.core.series.Series

This will return a Series. To extract a column as a DataFrame, you need to pass a list of column names. In our case that's just a single column:

In [84]:
genre_col = movies_df[["genre"]]
genre_col.head(5)

genre
Title                                            
Guardians of the Galaxy   Action,Adventure,Sci-Fi
Prometheus               Adventure,Mystery,Sci-Fi
Split                             Horror,Thriller
Sing                      Animation,Comedy,Family
Suicide Squad            Action,Adventure,Fantasy

In [85]:
type(genre_col)

pandas.core.frame.DataFrame

Since it's just a list, adding another column name is easy:

In [89]:
subset = movies_df[["genre", "rating"]]
subset.head()

genre  rating
Title                                                    
Guardians of the Galaxy   Action,Adventure,Sci-Fi     8.1
Prometheus               Adventure,Mystery,Sci-Fi     7.0
Split                             Horror,Thriller     7.3
Sing                      Animation,Comedy,Family     7.2
Suicide Squad            Action,Adventure,Fantasy     6.2

Now we'll look at getting data by rows.

__By Rows__

For rows, we have two options:

* `.loc` - locates by name
* `.iloc`- locates by numerical index

Remember that we are still indexed by movie Title, so to use `.loc` we give it the Title of a movie:

In [92]:
prom = movies_df.loc["Prometheus"]
prom

rank                                                                2
genre                                        Adventure,Mystery,Sci-Fi
description         Following clues to the origin of mankind, a te...
director                                                 Ridley Scott
actors              Noomi Rapace, Logan Marshall-Green, Michael Fa...
year                                                             2012
runtime                                                           124
rating                                                              7
votes                                                          485820
revenue_millions                                               126.46
metascore                                                          65
Name: Prometheus, dtype: object

On the other hand, with `iloc` we give it the numerical index of Prometheus:

In [94]:
prom = movies_df.iloc[1]
prom

rank                                                                2
genre                                        Adventure,Mystery,Sci-Fi
description         Following clues to the origin of mankind, a te...
director                                                 Ridley Scott
actors              Noomi Rapace, Logan Marshall-Green, Michael Fa...
year                                                             2012
runtime                                                           124
rating                                                              7
votes                                                          485820
revenue_millions                                               126.46
metascore                                                          65
Name: Prometheus, dtype: object

`loc` and `iloc` can be thought of as similar to Python list slicing. To show this even further, let's select multiple rows.

In [101]:
movie_subset = movies_df.loc["Prometheus": "Sing"]

movie_subset = movies_df.iloc[1:4]

#Both os these give same output

movie_subset

rank                     genre  \
Title                                        
Prometheus     2  Adventure,Mystery,Sci-Fi   
Split          3           Horror,Thriller   
Sing           4   Animation,Comedy,Family   

                                                  description  \
Title                                                           
Prometheus  Following clues to the origin of mankind, a te...   
Split       Three girls are kidnapped by a man with a diag...   
Sing        In a city of humanoid animals, a hustling thea...   

                        director  \
Title                              
Prometheus          Ridley Scott   
Split         M. Night Shyamalan   
Sing        Christophe Lourdelet   

                                                       actors  year  runtime  \
Title                                                                          
Prometheus  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012      124   
Split       James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016      117   
Sing        Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016      108   

            rating   votes  revenue_millions  metascore  
Title                                                    
Prometheus     7.0  485820            126.46       65.0  
Split          7.3  157606            138.12       62.0  
Sing           7.2   60545            270.32       59.0

One important distinction between using `.loc` and `.iloc` to select multiple rows is that `.loc` includes the movie _Sing_ in the result, but when using `.iloc` we're getting rows 1:4 but the movie at index 4 _(Suicide Squad)_ is not included.

Slicing with `.iloc` follows the same rules as slicing with lists, the object at the index at the end is not included.

#### Conditional selections

We’ve gone over how to select columns and rows, but what if we want to make a conditional selection?

For example, what if we want to filter our movies DataFrame to show only films directed by Ridley Scott or films with a rating greater than or equal to 8.0?

To do that, we take a column from the DataFrame and apply a Boolean condition to it. Here's an example of a Boolean condition:

In [103]:
condition = (movies_df["director"] == "Ridley Scott")
condition

Title
Guardians of the Galaxy    False
Prometheus                  True
Split                      False
Sing                       False
Suicide Squad              False
                           ...  
Secret in Their Eyes       False
Hostel: Part II            False
Step Up 2: The Streets     False
Search Party               False
Nine Lives                 False
Name: director, Length: 1000, dtype: bool

Similar to isnull(), this returns a Series of True and False values: True for films directed by Ridley Scott and False for ones not directed by him.

We want to filter out all movies not directed by Ridley Scott, in other words, we don’t want the False films. __To return the rows where that condition is True we have to pass this operation into the DataFrame:__

In [107]:
movies_df[movies_df["director"] == "Ridley Scott"]

#These condition can be read like:
# **  Select movies_df where movies_df director equals Ridley Scott **

rank                     genre  \
Title                                                    
Prometheus                 2  Adventure,Mystery,Sci-Fi   
The Martian              103    Adventure,Drama,Sci-Fi   
Robin Hood               388    Action,Adventure,Drama   
American Gangster        471     Biography,Crime,Drama   
Exodus: Gods and Kings   517    Action,Adventure,Drama   
The Counselor            522      Crime,Drama,Thriller   
A Good Year              531      Comedy,Drama,Romance   
Body of Lies             738      Action,Drama,Romance   

                                                              description  \
Title                                                                       
Prometheus              Following clues to the origin of mankind, a te...   
The Martian             An astronaut becomes stranded on Mars after hi...   
Robin Hood              In 12th century England, Robin and his band of...   
American Gangster       In 1970s America, a detective works to bring d...   
Exodus: Gods and Kings  The defiant leader Moses rises up against the ...   
The Counselor           A lawyer finds himself in over his head when h...   
A Good Year             A British investment broker inherits his uncle...   
Body of Lies            A CIA agent on the ground in Jordan hunts down...   

                            director  \
Title                                  
Prometheus              Ridley Scott   
The Martian             Ridley Scott   
Robin Hood              Ridley Scott   
American Gangster       Ridley Scott   
Exodus: Gods and Kings  Ridley Scott   
The Counselor           Ridley Scott   
A Good Year             Ridley Scott   
Body of Lies            Ridley Scott   

                                                                   actors  \
Title                                                                       
Prometheus              Noomi Rapace, Logan Marshall-Green, Michael Fa...   
The Martian             Matt Damon, Jessica Chastain, Kristen Wiig, Ka...   
Robin Hood              Russell Crowe, Cate Blanchett, Matthew Macfady...   
American Gangster       Denzel Washington, Russell Crowe, Chiwetel Eji...   
Exodus: Gods and Kings  Christian Bale, Joel Edgerton, Ben Kingsley, S...   
The Counselor           Michael Fassbender, Penélope Cruz, Cameron Dia...   
A Good Year             Russell Crowe, Abbie Cornish, Albert Finney, M...   
Body of Lies            Leonardo DiCaprio, Russell Crowe, Mark Strong,...   

                        year  runtime  rating   votes  revenue_millions  \
Title                                                                     
Prometheus              2012      124     7.0  485820            126.46   
The Martian             2015      144     8.0  556097            228.43   
Robin Hood              2010      140     6.7  221117            105.22   
American Gangster       2007      157     7.8  337835            130.13   
Exodus: Gods and Kings  2014      150     6.0  137299             65.01   
The Counselor           2013      117     5.3   84927             16.97   
A Good Year             2006      117     6.9   74674              7.46   
Body of Lies            2008      128     7.1  182305             39.38   

                        metascore  
Title                              
Prometheus                   65.0  
The Martian                  80.0  
Robin Hood                   53.0  
American Gangster            76.0  
Exodus: Gods and Kings       52.0  
The Counselor                48.0  
A Good Year                  47.0  
Body of Lies                 57.0

Let's look at conditional selections using numerical values by filtering the DataFrame by ratings:

In [109]:
movies_df[movies_df["rating"] >= 8.6].head(3)

rank                    genre  \
Title                                            
Interstellar       37   Adventure,Drama,Sci-Fi   
The Dark Knight    55       Action,Crime,Drama   
Inception          81  Action,Adventure,Sci-Fi   

                                                       description  \
Title                                                                
Interstellar     A team of explorers travel through a wormhole ...   
The Dark Knight  When the menace known as the Joker wreaks havo...   
Inception        A thief, who steals corporate secrets through ...   

                          director  \
Title                                
Interstellar     Christopher Nolan   
The Dark Knight  Christopher Nolan   
Inception        Christopher Nolan   

                                                            actors  year  \
Title                                                                      
Interstellar     Matthew McConaughey, Anne Hathaway, Jessica Ch...  2014   
The Dark Knight  Christian Bale, Heath Ledger, Aaron Eckhart,Mi...  2008   
Inception        Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...  2010   

                 runtime  rating    votes  revenue_millions  metascore  
Title                                                                   
Interstellar         169     8.6  1047747            187.99       74.0  
The Dark Knight      152     9.0  1791916            533.32       82.0  
Inception            148     8.8  1583625            292.57       74.0

We can make some richer conditionals by using logical operators | for "or" and & for "and".

Let's filter the the DataFrame to show only movies by Christopher Nolan OR Ridley Scott:

In [111]:
movies_df[(movies_df["director"] == "Christopher Nolan") | (movies_df["director"] == " Ridley Scott")]


rank                    genre  \
Title                                                  
Interstellar             37   Adventure,Drama,Sci-Fi   
The Dark Knight          55       Action,Crime,Drama   
The Prestige             65     Drama,Mystery,Sci-Fi   
Inception                81  Action,Adventure,Sci-Fi   
The Dark Knight Rises   125          Action,Thriller   

                                                             description  \
Title                                                                      
Interstellar           A team of explorers travel through a wormhole ...   
The Dark Knight        When the menace known as the Joker wreaks havo...   
The Prestige           Two stage magicians engage in competitive one-...   
Inception              A thief, who steals corporate secrets through ...   
The Dark Knight Rises  Eight years after the Joker's reign of anarchy...   

                                director  \
Title                                      
Interstellar           Christopher Nolan   
The Dark Knight        Christopher Nolan   
The Prestige           Christopher Nolan   
Inception              Christopher Nolan   
The Dark Knight Rises  Christopher Nolan   

                                                                  actors  \
Title                                                                      
Interstellar           Matthew McConaughey, Anne Hathaway, Jessica Ch...   
The Dark Knight        Christian Bale, Heath Ledger, Aaron Eckhart,Mi...   
The Prestige           Christian Bale, Hugh Jackman, Scarlett Johanss...   
Inception              Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...   
The Dark Knight Rises  Christian Bale, Tom Hardy, Anne Hathaway,Gary ...   

                       year  runtime  rating    votes  revenue_millions  \
Title                                                                     
Interstellar           2014      169     8.6  1047747            187.99   
The Dark Knight        2008      152     9.0  1791916            533.32   
The Prestige           2006      130     8.5   913152             53.08   
Inception              2010      148     8.8  1583625            292.57   
The Dark Knight Rises  2012      164     8.5  1222645            448.13   

                       metascore  
Title                             
Interstellar                74.0  
The Dark Knight             82.0  
The Prestige                66.0  
Inception                   74.0  
The Dark Knight Rises       78.0

movies_df [ (condition 1) | (condition 2) ]

We need to make sure to group evaluations with parentheses so Python knows how to evaluate the conditional.

Using the `isin()` method we could make this more concise though:

In [115]:
movies_df[movies_df["director"].isin(["Christopher Nolan", "Ridley Scott"])].head(2)

rank                     genre  \
Title                                          
Prometheus       2  Adventure,Mystery,Sci-Fi   
Interstellar    37    Adventure,Drama,Sci-Fi   

                                                    description  \
Title                                                             
Prometheus    Following clues to the origin of mankind, a te...   
Interstellar  A team of explorers travel through a wormhole ...   

                       director  \
Title                             
Prometheus         Ridley Scott   
Interstellar  Christopher Nolan   

                                                         actors  year  \
Title                                                                   
Prometheus    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
Interstellar  Matthew McConaughey, Anne Hathaway, Jessica Ch...  2014   

              runtime  rating    votes  revenue_millions  metascore  
Title                                                                
Prometheus        124     7.0   485820            126.46       65.0  
Interstellar      169     8.6  1047747            187.99       74.0

Let's say we want all movies that were released between 2005 and 2010, have a rating above 8.0, but made below the 25th percentile in revenue.

Here's how we could do all of that:

In [121]:
movies_df[
           ((movies_df["year"] >= 2005) & (movies_df["year"] <= 2010))
            & (movies_df["rating"] > 8.0)
            & (movies_df["revenue_millions"] < movies_df["revenue_millions"].quantile(0.25))
              ]

rank               genre  \
Title                                           
3 Idiots              431        Comedy,Drama   
The Lives of Others   477      Drama,Thriller   
Incendies             714   Drama,Mystery,War   
Taare Zameen Par      992  Drama,Family,Music   

                                                           description  \
Title                                                                    
3 Idiots             Two friends are searching for their long lost ...   
The Lives of Others  In 1984 East Berlin, an agent of the secret po...   
Incendies            Twins journey to the Middle East to discover t...   
Taare Zameen Par     An eight-year-old boy is thought to be a lazy ...   

                                             director  \
Title                                                   
3 Idiots                              Rajkumar Hirani   
The Lives of Others  Florian Henckel von Donnersmarck   
Incendies                            Denis Villeneuve   
Taare Zameen Par                           Aamir Khan   

                                                                actors  year  \
Title                                                                          
3 Idiots               Aamir Khan, Madhavan, Mona Singh, Sharman Joshi  2009   
The Lives of Others  Ulrich Mühe, Martina Gedeck,Sebastian Koch, Ul...  2006   
Incendies            Lubna Azabal, Mélissa Désormeaux-Poulin, Maxim...  2010   
Taare Zameen Par     Darsheel Safary, Aamir Khan, Tanay Chheda, Sac...  2007   

                     runtime  rating   votes  revenue_millions  metascore  
Title                                                                      
3 Idiots                 170     8.4  238789              6.52       67.0  
The Lives of Others      137     8.5  278103             11.28       89.0  
Incendies                131     8.2   92863              6.86       80.0  
Taare Zameen Par         165     8.5  102697              1.20       42.0

If you recall up when we used __.describe()__ the 25th percentile for revenue was about 17.4, and we can access this value directly by using the __quantile()__ method with a float of 0.25.

### Applying functions

It is possible to iterate over a DataFrame or Series as you would with a list, but doing so — especially on large datasets — is very slow.

An efficient alternative is to `apply()` a function to the dataset. For example, we could use a function to convert movies with an 8.0 or greater to a string value of "good" and the rest to "bad" and use this transformed values to create a new column.

First we would create a function that, when given a rating, determines if it's good or bad:

In [122]:
def rating_function(x):
    if x >= 8.0:
        return "good"
    else:
        return "bad"
    

Now we want to send the entire rating column through this function, which is what __apply()__ does:

In [128]:
movies_df["rating_category"] = movies_df["rating"].apply(rating_function)
movies_df.head(2)

rank                     genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   

                                                               description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   

                             director  \
Title                                   
Guardians of the Galaxy    James Gunn   
Prometheus               Ridley Scott   

                                                                    actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   

                         year  runtime  rating   votes  revenue_millions  \
Title                                                                      
Guardians of the Galaxy  2014      121     8.1  757074            333.13   
Prometheus               2012      124     7.0  485820            126.46   

                         metascore rating_category  
Title                                               
Guardians of the Galaxy       76.0            good  
Prometheus                    65.0             bad

The `.apply()` method passes every value in the rating column through the __rating_function__ and then returns a new Series. This Series is then assigned to a new column called __rating_category.__

In [126]:
movies_df["rating_category"] = movies_df["rating"].apply(lambda x: "good" if x >8.0 else "bad") 
movies_df.head(2)

rank                     genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   

                                                               description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   

                             director  \
Title                                   
Guardians of the Galaxy    James Gunn   
Prometheus               Ridley Scott   

                                                                    actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   

                         year  runtime  rating   votes  revenue_millions  \
Title                                                                      
Guardians of the Galaxy  2014      121     8.1  757074            333.13   
Prometheus               2012      124     7.0  485820            126.46   

                         metascore rating_category  
Title                                               
Guardians of the Galaxy       76.0            good  
Prometheus                    65.0             bad

Overall, using `apply()` will be much faster than iterating manually over rows because pandas is utilizing vectorization.